src/test/java/org/synyx/urlaubsverwaltung/statistics/web/ApplicationForLeaveStatisticsControllerTest.java

```diff
@@ -0,0 +1,166 @@
+import java.util.Optional;
+
+import static org.hamcrest.beans.SamePropertyValuesAs.samePropertyValuesAs;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.ArgumentMatchers.eq;
+import static org.mockito.ArgumentMatchers.refEq;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
+import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.header;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.model;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.view;
+import static org.springframework.test.web.servlet.setup.MockMvcBuilders.standaloneSetup;
+
+@RunWith(MockitoJUnitRunner.class)
+public class ApplicationForLeaveStatisticsControllerTest {
+
+    private ApplicationForLeaveStatisticsController sut;
+
+    @Mock
+    private ApplicationForLeaveStatisticsService applicationForLeaveStatisticsService;
+
+    @Mock
+    private ApplicationForLeaveStatisticsCsvExportService applicationForLeaveStatisticsCsvExportService;
+
+    @Mock
+    private VacationTypeService vacationTypeService;
+
+    @Before
+    public void setUp() {
+
+        sut = new ApplicationForLeaveStatisticsController(applicationForLeaveStatisticsService, applicationForLeaveStatisticsCsvExportService, vacationTypeService);
+    }
+
+    @Test
+    public void applicationForLeaveStatisticsRedirectsToStatistics() throws Exception {
+
+        final FilterPeriod filterPeriod = new FilterPeriod();
+        final String expectedRedirect = "/web/application/statistics?from=" + filterPeriod.getStartDateAsString() +
+                                        "&to=" + filterPeriod.getEndDateAsString();
+
+        perform(post("/web/application/statistics"))
+            .andExpect(status().isFound())
+            .andExpect(header().string("Location", expectedRedirect));
+    }
+
+    @Test
+    public void applicationForLeaveStatisticsAddsErrorToModelAndShowsFormIfPeriodNotTheSameYear() throws Exception {
+
+        final String startDate = "01.01.2000";
+        final String endDate = "01.01.2019";
+
+        perform(get("/web/application/statistics")
+            .param("from", startDate)
+            .param("to", endDate))
+            .andExpect(model().attribute("errors", "INVALID_PERIOD"))
+            .andExpect(view().name("application/app_statistics"));
+    }
+
+    @Test
+    public void applicationForLeaveStatisticsSetsModelAndView() throws Exception {
+
+        final String startDate = "01.01.2019";
+        final String endDate = "01.08.2019";
+
+        final FilterPeriod period = new FilterPeriod(Optional.ofNullable(startDate), Optional.ofNullable(endDate));
```

Comment:
hat es einen grund für `Optional.ofNullable` anstatt `Optional.of`?
wir wissen hier ja, dass der wert != null ist und können `.of` nehmen, das ist kürzer :-)

Reply:
Du hast Recht, habe angepasst ;)